notebooks/Binary Classification with the UCI Credit Card Payment Default Dataset.ipynb

```diff
@@ -0,0 +1,599 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We calculate several disparity metrics below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Metrics\n",
+    "models_dict = {\"Unmitigated\": (test_preds, test_scores)}\n",
+    "get_metrics_df(models_dict, Y_test, A_str_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Mitigating Demographic Disparities with Postprocessing Techniques"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We attempt to mitigate the disparities in the `lightgbm` prediction using the `fairlearn` postprocessing technique `ThersholdOptimizer`. This technique will find a suitable threshold for the scores (class probabilities) produced by the `lightgbm` model such that the demographic disparities are minimized. For this exercise, we focus on the `equalized odds` disparity metric."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "postprocess_est = ThresholdOptimizer(\n",
+    "    estimator=model,\n",
+    "    constraints=\"equalized_odds\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "balanced_idx1 = df_train[Y_train==1].index\n",
+    "pp_train_idx = balanced_idx1.union(Y_train[Y_train==0].sample(n=balanced_idx1.size, random_state=1234).index)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_train_balanced = df_train.loc[pp_train_idx, :]\n",
+    "Y_train_balanced = Y_train.loc[pp_train_idx]\n",
+    "A_train_balanced = A_train.loc[pp_train_idx]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "postprocess_est.fit(df_train_balanced, Y_train_balanced, sensitive_features=A_train_balanced)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "postprocess_preds = postprocess_est.predict(df_test, sensitive_features=A_test)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "models_dict = {\"Unmitigated\": (test_preds, test_scores),\n",
+    "              \"ThresholdOptimizer\": (postprocess_preds, postprocess_preds)}\n",
+    "get_metrics_df(models_dict, Y_test, A_str_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We note that the `ThresholdOptimizer` method significantly reduces demographic disparity across multiple metrics. \n",
+    "\n",
+    "Below, we compare this model with the unmitigated `lightgbm` model using the `fairlearn` dashboard. "
```

Comment:
Consider suggesting things for the user to examine with the dashboard.

Reply:
addressed.